### ISSUE TITLE
Mutate a file on Windows doesn't work

### ISSUE THREAD

**Issue by alexchapar** [slcp/stryker-runner#183]
I stumble around an issue on Windows trying to mutate a file. I kinda draft a possible solution (don't know if it could create issue on other OS though 😉). If you want me to do a PR about it, just let me know 🙂 

## Description of the problem

Running the mutation on a file (with or without a selection) on Windows does not work.

The generated command line looks like this:

```powershell
npx --no-install stryker run --mutate /d:/source/github/org/project/src/file.ts ./stryker.conf.js
```

And some errors are displayed:

```powershell
00:14:28 (33452) WARN ProjectReader Glob pattern "/d:/source/github/org/project/src/file.ts" did not result in any files.
00:14:28 (33452) WARN ProjectReader No files marked to be mutated, Stryker will perform a dry-run without actually mutating anything. You can configure the `mutate` property in your config file (or use `--mutate` via command line).
```

There is a couple issues/explanations around this:

- The first problem, is that is tries to mutate a file for the path starting with a slash /c:/.../file.ts. Windows doesn't use the slash, it starts with the "drive letter" 😉 
- The second one is that it seems to missing some quotes around the file... I don't know exactly why. Based on the stryker doc it should be ok for one specific file... But still wasn't working for me until I added quotes around the file (including the selection range if exists)
- The third and last one is about the full file path which stryker-js seems to have issue to retrieve the file to mutate (again, it may only be on windows).

## Expected behaviour

So, for the Windows folks...

We would expect that the command generated to mutate a file (with or without a selection) would be one of the two :
- `[yarn or npx --no-install] stryker run --mutate "src\file-to-mutate.ts:13-25" ./stryker.conf.js`
- `[yarn or npx --no-install] stryker run --mutate "**\file-to-mutate.ts" ./stryker.conf.js` (but not possible with a selection though)

## Possible fix

- The use of "file.fsPath", instead of file.path, to assign the target within the makeCommand of the stryker.js should help. The use of `fsPath` should return the adequate "fullpath" based on the file system. (it should resolve the first problem)
- Adding quotes around the `\"${target}\"` while creating the stryker command to return should resolve the issue about the file (ie the second problem listen earlier)
- The trickier behaviour (ie third problem) is about the filepath to use. Should it be relative to the workspace ? Or Should it be using some wildcard ? Either seems to work for files.
  - The [wildcard](#1-wildcard) is easier to fix but may create a little problem if multiple files has the same name which shouldn't happen much, but still... Other downside, it won't work on a selection 😞 
  - The use of the [relative path](#2-relative-path) is a bit less fun to code and test (and covering mutations). but will ensure that only one file will be mutated and the selection will work without any changes.

### 1-Wildcard

A wildcard solution could be this (but still require something different for the selection one)

```typescript
  import path from 'path';
  //...
  const makeCommand = //...
  //...
    const target = `${path.basename(file.fsPath)}${lineRange ? `:${lineRange}` : ''}`;
    // TODO : find a way to change to relative path when the line range are required
    return `${strykerBin} run --mutate \"**\\${target}\"${configFilePath ? ` ${configFilePath}` : ''}`;
```

### 2-Relative Path

A probable solution using a relative path could be something like this

```typescript
  //...
  const makeCommand = //...
  //...
    const target = `${path.relative(vscode.workspace.getWorkspaceFolder(file)?.uri?.fsPath!, file.fsPath)}${lineRange ? `:${lineRange}` : ''}`;
    return `${strykerBin} run --mutate \"${target}\"${configFilePath ? ` ${configFilePath}` : ''}`;
```

## Versions/OS

- vscode v1.77 & 1.78
- stryker-runner v0.0.40 & 0.0.41
- windows 10

**Comment by slcp** [comment]
Thanks for raising this. The daily usage of this that I have seen has all been unix systems so getting some input using it cross platform is very welcome :) Once the problem is fully understood a PR would also be welcome. Addressing the 3 issues identified:

1. If I am understanding the issues correctly the use of `fsPath` property of  a [URI](https://code.visualstudio.com/api/references/vscode-api#Uri) seems like the a good solution and from reading the docs it should work cross platform.
2. Also, adding quotes (`"`) around the Stryker command should be no issue cross platform.
3. I'm not fully clear on this one, do you mean that Stryker on Windows appears to only be able to resolve relative paths and never an absolute path?

**Comment by alexchapar** [comment]
Yes, for the third one, it looks a bit like it... [Disclaimer: Although I only tested on Typescript 🙄]

The documentation for stryker-js is not very accurate either. There is only one mention of the relative path when using "--ignorePatterns". Also, any mention of "--mutate" is used with "src/*" (or similar) which, we could argue, "may" (or not) be relative 😉. I haven't found any "github issues" about this either on stryker-js.

Assuming my source project's root path is "c:\dev\project" (and thus my workspace and terminal path use that folder as a starting point), and using the command line generated from stryker-runner and adjusting it afterwards to reduce the problem :

These did not work:

- npx --no-install stryker run --mutate "/c:/dev/project/src/file.ts" ./stryker.conf.js
- npx --no-install stryker run --mutate "c:/dev/project/src/file.ts" ./stryker.conf.js
- npx --no-install stryker run --mutate "/src/file.ts" ./stryker.conf.js
- npx --no-install stryker run --mutate "/src/file.ts" ./stryker.conf.js
- npx --no-install stryker run --mutate "file.ts" ./stryker.conf.js

and gave the following warnings:

> 21:27:03 (33076) WARN ProjectReader Glob pattern "`THE --MUTATE FILE IN COMAND LINE`" did not result in any files.
> 21:27:03 (33076) WARN ProjectReader No files marked to be mutated, Stryker will perform a dry-run without actually mutating anything. You can configure the `mutate` property in your config file (or use `--mutate` via command line).

These have worked:

- npx --no-install stryker run --mutate "./src/file.ts" ./stryker.conf.js
- npx --no-install stryker run --mutate "src/file.ts" ./stryker.conf.js
- npx --no-install stryker run --mutate "src\file.ts" ./stryker.conf.js
- npx --no-install stryker run --mutate "src/**/*.ts" ./stryker.conf.js
- npx --no-install stryker run --mutate "**/file.ts" ./stryker.conf.js

and found a file to mutate:

> 21:27:14 (22816) INFO ProjectReader Found 1 of 40 file(s) to be mutated.
> 21:27:15 (22816) INFO Instrumenter Instrumented 1 source file(s) with 16 mutant(s)

Also, it's not a problem to use backslashes or slashes, since Windows can understand both. And the mutation of a range reacted similarly as the file.

The funny thing about quotation marks is that they may not be necessary after all... Or at least I'm no longer able to reproduce the issue I thought I had 😂. So, the use of quotes, if retained, might be interesting if one wanted to, possibly, allow more files to be mutated at the same time (as explained in the [mutate section of the configuration](https://stryker-mutator.io/docs/stryker-js/configuration/#mutate-string) page) - maybe for a future feature! 😆.

**Comment by alexchapar** [comment]
I created a draft PR, to give you some ideas... it works on Windows, but don't know if it's going to pass the tests on unix... I will try to spin a linux somewhere to test it out. But if you have the time to test it before me, let me know... 

Also, if you have any comments on what I started in the PR don't hesitate. (also I did a little mistake within the PR and it also contains the modif from the previous PR... it will be corrected before enabling "Ready for review")

**Comment by alexchapar** [comment]
It took me a while, but I was finally able to find the time to validate that the "Pull Request" concerning the use of the "fsPath" and the relative path also work on Ubuntu, using VSCode and VSCodium. 

However, I'm not able to validate/test on a MacOS...

**Comment by slcp** [comment]
Closed with #192
